### ISSUE TITLE
Warn about possible performance issue (>>) and (*>)

### ISSUE THREAD

**Issue by alanz** [hvr/Hs2010To201x#1]
In https://mail.haskell.org/pipermail/libraries/2015-November/026504.html @takano-akio suggests

> I understand this, but perhaps there is a way to achieve this without
> slowing down existing code. How about introducing a new warning
> (enabled with -Wall) that is triggered when a type satisfies the
> following 3 conditions?
> 1. The type has a Monad instance and an Applicative instance declared
>    in the same module, with the same set of constraints.
> 2. `(*>)` is not defined as `(*>)` = `(>>)`. i.e. either it has a
>    non-trivial definition or its definition is left out.
> 3. `(>>)` is not defined as `(>>)` = `(*>)`. i.e. either it has a
>    non-trivial definition or its definition is left out.
> 
> This way, people can be warned when `(*>)` and `(>>)` can share an
> implementation but they don't.

**Comment by hvr** [comment]
@alanz you may want to `>`-quote the description otherwise it looks like you were the author rather than @takano-akio :-)

related: https://github.com/ghc/ghc/commit/f09f2470a76bb08b7f51d2f5663daa672b86f618

**Comment by alanz** [comment]
Yes, as I did that I realised that I should have put a link to the mail in.
Will update it.

**Comment by hvr** [comment]
Btw, the suggestion as-is has one obvious flaw as it suggests to define `(*>)` and `(>>)` mutually in terms of each other. In contrast, the warnings I implemented in GHC are more robust and much simpler to implement and don't need to access typechecking information (e.g. for testing existence of instances for a type).

**Comment by alanz** [comment]
I just captured it as an issue so it doesn't get lost. If it is done in GHC
then so much the better, it will warn always.

**Comment by hvr** [comment]
except those warnings aren't enabled by default yet, because I'm not sure if that's desired by the larger community... I'm currently also still testing it on Hackage
